contracts/azorius/strategies/HatsProposalCreationWhitelistV1.sol

```diff
@@ -0,0 +1,115 @@
+    }
+
+    /**
+     * Sets up the contract with its initial parameters.
+     *
+     * @param initializeParams encoded initialization parameters:
+     * `address _hatsContract`, `uint256[] _initialWhitelistedHats`
+     */
+    function setUp(bytes memory initializeParams) public virtual {
+        (address _hatsContract, uint256[] memory _initialWhitelistedHats) = abi
+            .decode(initializeParams, (address, uint256[]));
+
+        if (_hatsContract == address(0)) revert MissingHatsContract();
+        hatsContract = IHats(_hatsContract);
+
+        if (_initialWhitelistedHats.length == 0) revert NoHatsWhitelisted();
+        for (uint256 i = 0; i < _initialWhitelistedHats.length; i++) {
+            _whitelistHat(_initialWhitelistedHats[i]);
+        }
+    }
+
+    /**
+     * Adds a Hat to the whitelist for proposal creation.
+     * @param _hatId The ID of the Hat to whitelist
+     */
+    function whitelistHat(uint256 _hatId) external onlyOwner {
+        _whitelistHat(_hatId);
+    }
+
+    /**
+     * Internal function to add a Hat to the whitelist.
+     * @param _hatId The ID of the Hat to whitelist
+     */
+    function _whitelistHat(uint256 _hatId) internal {
+        for (uint256 i = 0; i < whitelistedHatIds.length; i++) {
+            if (whitelistedHatIds[i] == _hatId) revert HatAlreadyWhitelisted();
+        }
+        whitelistedHatIds.push(_hatId);
+        emit HatWhitelisted(_hatId);
+    }
+
+    /**
+     * Removes a Hat from the whitelist for proposal creation.
+     * @param _hatId The ID of the Hat to remove from the whitelist
+     */
+    function unwhitelistHat(uint256 _hatId) external onlyOwner {
+        bool found = false;
+        for (uint256 i = 0; i < whitelistedHatIds.length; i++) {
+            if (whitelistedHatIds[i] == _hatId) {
+                whitelistedHatIds[i] = whitelistedHatIds[
+                    whitelistedHatIds.length - 1
+                ];
+                whitelistedHatIds.pop();
+                found = true;
+                break;
+            }
+        }
+        if (!found) revert HatNotWhitelisted();
+
+        emit HatUnwhitelisted(_hatId);
+    }
+
+    /**
+     * @dev Checks if an address is authorized to create proposals.
+     * @param _address The address to check for proposal creation authorization.
+     * @return bool Returns true if the address is wearing any of the whitelisted Hats, false otherwise.
+     * @notice This function overrides the isProposer function from the parent contract.
+     * It iterates through all whitelisted Hat IDs and checks if the given address
+     * is wearing any of them using the Hats Protocol.
+     */
+    function isProposer(address _address) public view virtual returns (bool) {
+        for (uint256 i = 0; i < whitelistedHatIds.length; i++) {
+            if (hatsContract.isWearerOfHat(_address, whitelistedHatIds[i])) {
+                return true;
+            }
+        }
+        return false;
+    }
```

Comment:
should we also be doing other checks here for standing of that hat? For example there are functions like:

https://github.com/Hats-Protocol/hats-protocol/blob/main/src/Hats.sol
- `checkHatWearerStatus` Line 372
- `isActive` Line 1042
- `isInGoodStanding` Line 1066

Or wearer being all we are looking for in this system?

Reply:
- `checkHatWearerStatus` is not a view function, it can update state. so I don't want to call it in this `isProposer` function, which must be a view function
- `isWearerOfHat` calls `balanceOf` which checks both `isActive` and `isEligible`
- `isInGoodStanding` is interesting... I'm honestly not sure about this. Looking at the implementation, it calls a function on the given hat's `eligibility` module. We don't use `eligibility` modules in our system (afaik), so I don't think this is really relevant for us. I've only put a few minutes of thought into this at this point though.

Reply:
- makes sense.

For the other two. yeah. I'm down to talk about this further.